File: pixie/dag/trace_check.py
"""Validate a captured trace tree against a data-flow DAG."""

from __future__ import annotations

from dataclasses import dataclass, field
from pathlib import Path
from typing import Any

from pixie.dag import DagNode, is_valid_dag_name, parse_dag


@dataclass
class TraceCheckResult:
    """Result of checking a trace against the DAG."""

    valid: bool
    matched: list[str] = field(default_factory=list)  # DAG node names that matched
    unmatched: list[str] = field(
        default_factory=list
    )  # DAG node names not found in trace
    extra_spans: list[str] = field(default_factory=list)  # span names not in DAG
    errors: list[str] = field(default_factory=list)


def _collect_span_info(
    nodes: list[Any],
) -> list[dict[str, str]]:
    """Recursively collect span name and type info from a trace tree."""
    from pixie.instrumentation.spans import LLMSpan

    info: list[dict[str, str]] = []
    for node in nodes:
        span_type = "llm_call" if isinstance(node.span, LLMSpan) else "observation"
        info.append({"name": node.name, "type": span_type})
        info.extend(_collect_span_info(node.children))
    return info


def check_trace_against_dag(
    dag_nodes: list[DagNode],
    trace_tree: list[Any],
) -> TraceCheckResult:
    """Check that the trace tree contains spans matching the DAG nodes.

    Matching rules:

    - If ``is_llm_call`` is ``true`` on a DAG node, the check passes when at
      least one LLM span exists in the trace. Name matching is skipped.
    - Otherwise, the DAG node name must match a non-LLM span name exactly.
    """
    result = TraceCheckResult(valid=True)
    unmatched_reasons: dict[str, str] = {}

    # Enforce naming contract even when callers skip explicit DAG validation.
    invalid_names: list[str] = []
    for node in dag_nodes:
        if not is_valid_dag_name(node.name):
            invalid_names.append(
                f"DAG node '{node.name}': name must be lower_snake_case "
                "(e.g., 'handle_turn')."
            )
        if node.parent is not None and not is_valid_dag_name(node.parent):
            invalid_names.append(
                f"DAG node '{node.name}': parent '{node.parent}' must be "
                "lower_snake_case."
            )
    if invalid_names:
        return TraceCheckResult(valid=False, errors=invalid_names)

    # Collect spans from trace
    span_info = _collect_span_info(trace_tree)
    span_names_by_type: dict[str, set[str]] = {"observation": set(), "llm_call": set()}
    for span in span_info:
        span_names_by_type.setdefault(span["type"], set()).add(span["name"])

    span_names = span_names_by_type["observation"] | span_names_by_type["llm_call"]
    has_llm_spans = any(s["type"] == "llm_call" for s in span_info)
    has_llm_dag_nodes = any(node.is_llm_call for node in dag_nodes)

    # Check each DAG node has a matching span according to its llm flag.
    matched_span_names: set[str] = set()
    for dag_node in dag_nodes:
        if dag_node.is_llm_call:
            if has_llm_spans:
                result.matched.append(dag_node.name)
            else:
                result.unmatched.append(dag_node.name)
                unmatched_reasons[dag_node.name] = "missing_llm_span"
            continue

        if dag_node.name not in span_names:
            result.unmatched.append(dag_node.name)
            unmatched_reasons[dag_node.name] = "missing_named_span"
            continue

        # For non-LLM DAG nodes, ensure the matched span is non-LLM.
        if (
            dag_node.name in span_names_by_type["llm_call"]
            and dag_node.name not in span_names_by_type["observation"]
        ):
            result.unmatched.append(dag_node.name)
            unmatched_reasons[dag_node.name] = "llm_flag_mismatch"
        else:
            result.matched.append(dag_node.name)
            matched_span_names.add(dag_node.name)

    # Find spans not accounted for by the DAG
    for span_name in sorted(span_names_by_type["observation"]):
        if span_name not in matched_span_names:
            result.extra_spans.append(span_name)
    if not has_llm_dag_nodes:
        for span_name in sorted(span_names_by_type["llm_call"]):
            result.extra_spans.append(span_name)

    if result.unmatched:
        result.valid = False
        for node_name in result.unmatched:
            node = next(n for n in dag_nodes if n.name == node_name)
            reason = unmatched_reasons.get(node_name)
            if reason == "missing_llm_span":
                result.errors.append(
                    f"DAG node '{node.name}' (is_llm_call=true) expects at least "
                    f"one LLM span in the trace, but none were found. "
                    f"Common fix: ensure `enable_storage()` is called BEFORE the "
                    f"LLM client (OpenAI, Anthropic, etc.) is created."
                )
            elif reason == "llm_flag_mismatch":
                result.errors.append(
                    f"DAG node '{node.name}' has is_llm_call=false but matched an LLM "
                    "span. Fix: set `is_llm_call: true` on this node in the DAG JSON."
                )
            else:
                result.errors.append(
                    f"DAG node '{node.name}' has no matching span in the trace. "
                    f"This means either: (1) the function at `code_pointer` is not "
                    f"decorated with `@observe(name='{node.name}')`, or (2) the "
                    f"function was not called during the trace run. Fix: add "
                    f"`@observe(name='{node.name}')` to the function specified by "
                    f"the node's `code_pointer`, or if already decorated, ensure "
                    f"the name matches exactly."
                )

    return result


async def check_last_trace(
    dag_json_path: Path,
) -> TraceCheckResult:
    """Load the most recent trace and check it against a DAG JSON file.

    Returns a TraceCheckResult with match details.
    """
    from piccolo.engine.sqlite import SQLiteEngine

    from pixie.config import get_config
    from pixie.storage.store import ObservationStore

    # Parse DAG
    dag_nodes, parse_errors = parse_dag(dag_json_path)
    if parse_errors:
        return TraceCheckResult(valid=False, errors=parse_errors)

    # Load latest trace
    config = get_config()
    engine = SQLiteEngine(path=config.db_path)
    store = ObservationStore(engine=engine)

    traces = await store.list_traces(limit=1)
    if not traces:
        return TraceCheckResult(
            valid=False,
            errors=["No traces found. Run the app first to produce a trace."],
        )

    trace_id = traces[0]["trace_id"]
    tree = await store.get_trace(trace_id)
    if not tree:
        return TraceCheckResult(
            valid=False,
            errors=[f"No spans found for trace '{trace_id}'."],
        )

    return check_trace_against_dag(dag_nodes, tree)

File: ts/src/dag/traceCheck.ts
/**
 * Validate a captured trace tree against a data-flow DAG.
 */

import type { ObservationNode } from "../storage/tree";
import type { DagNode } from "./index";
import { isValidDagName, parseDag } from "./index";

/** Result of checking a trace against the DAG. */
export interface TraceCheckResult {
  valid: boolean;
  /** DAG node names that matched. */
  matched: string[];
  /** DAG node names not found in trace. */
  unmatched: string[];
  /** Span names not in DAG. */
  extraSpans: string[];
  errors: string[];
}

/**
 * Recursively collect span name and type info from a trace tree.
 */
function collectSpanInfo(
  nodes: ObservationNode[]
): Array<{ name: string; type: string }> {
  const info: Array<{ name: string; type: string }> = [];
  for (const node of nodes) {
    const spanType = "operation" in node.span ? "llm_call" : "observation";
    info.push({ name: node.name, type: spanType });
    info.push(...collectSpanInfo(node.children));
  }
  return info;
}

/**
 * Check that the trace tree contains spans matching the DAG nodes.
 *
 * Matching rules:
 * - If `isLlmCall` is true on a DAG node, the check passes when at
 *   least one LLM span exists in the trace. Name matching is skipped.
 * - Otherwise, the DAG node name must match a non-LLM span name exactly.
 */
export function checkTraceAgainstDag(
  dagNodes: DagNode[],
  traceTree: ObservationNode[]
): TraceCheckResult {
  const result: TraceCheckResult = {
    valid: true,
    matched: [],
    unmatched: [],
    extraSpans: [],
    errors: [],
  };
  const unmatchedReasons = new Map<string, string>();

  // Enforce naming contract
  const invalidNames: string[] = [];
  for (const node of dagNodes) {
    if (!isValidDagName(node.name)) {
      invalidNames.push(
        `DAG node '${node.name}': name must be lower_snake_case (e.g., 'handle_turn').`
      );
    }
    if (node.parent !== null && !isValidDagName(node.parent)) {
      invalidNames.push(
        `DAG node '${node.name}': parent '${node.parent}' must be lower_snake_case.`
      );
    }
  }
  if (invalidNames.length > 0) {
    return { valid: false, matched: [], unmatched: [], extraSpans: [], errors: invalidNames };
  }

  // Collect spans from trace
  const spanInfo = collectSpanInfo(traceTree);
  const spanNamesByType: Record<string, Set<string>> = {
    observation: new Set<string>(),
    llm_call: new Set<string>(),
  };
  for (const span of spanInfo) {
    if (!spanNamesByType[span.type]) {
      spanNamesByType[span.type] = new Set();
    }
    spanNamesByType[span.type].add(span.name);
  }

  const spanNames = new Set([
    ...spanNamesByType["observation"],
    ...spanNamesByType["llm_call"],
  ]);
  const hasLlmSpans = spanInfo.some((s) => s.type === "llm_call");
  const hasLlmDagNodes = dagNodes.some((n) => n.isLlmCall);

  // Check each DAG node has a matching span
  const matchedSpanNames = new Set<string>();
  for (const dagNode of dagNodes) {
    if (dagNode.isLlmCall) {
      if (hasLlmSpans) {
        result.matched.push(dagNode.name);
      } else {
        result.unmatched.push(dagNode.name);
        unmatchedReasons.set(dagNode.name, "missing_llm_span");
      }
      continue;
    }

    if (!spanNames.has(dagNode.name)) {
      result.unmatched.push(dagNode.name);
      unmatchedReasons.set(dagNode.name, "missing_named_span");
      continue;
    }

    // For non-LLM DAG nodes, ensure the matched span is non-LLM
    if (
      spanNamesByType["llm_call"].has(dagNode.name) &&
      !spanNamesByType["observation"].has(dagNode.name)
    ) {
      result.unmatched.push(dagNode.name);
      unmatchedReasons.set(dagNode.name, "llm_flag_mismatch");
    } else {
      result.matched.push(dagNode.name);
      matchedSpanNames.add(dagNode.name);
    }
  }

  // Find spans not accounted for by the DAG
  for (const spanName of [...spanNamesByType["observation"]].sort()) {
    if (!matchedSpanNames.has(spanName)) {
      result.extraSpans.push(spanName);
    }
  }
  if (!hasLlmDagNodes) {
    for (const spanName of [...spanNamesByType["llm_call"]].sort()) {
      result.extraSpans.push(spanName);
    }
  }

  if (result.unmatched.length > 0) {
    result.valid = false;
    for (const nodeName of result.unmatched) {
      const node = dagNodes.find((n) => n.name === nodeName)!;
      const reason = unmatchedReasons.get(nodeName);
      if (reason === "missing_llm_span") {
        result.errors.push(
          `DAG node '${node.name}' (is_llm_call=true) expects at least ` +
          `one LLM span in the trace, but none were found. ` +
          `Common fix: ensure \`enableStorage()\` is called BEFORE the ` +
          `LLM client (OpenAI, Anthropic, etc.) is created.`
        );
      } else if (reason === "llm_flag_mismatch") {
        result.errors.push(
          `DAG node '${node.name}' has is_llm_call=false but matched an LLM ` +
          `span. Fix: set \`isLlmCall: true\` on this node in the DAG JSON.`
        );
      } else {
        result.errors.push(
          `DAG node '${node.name}' has no matching span in the trace. ` +
          `This means either: (1) the function at \`codePointer\` is not ` +
          `decorated with \`observe(fn, '${node.name}')\`, or (2) the ` +
          `function was not called during the trace run. Fix: wrap the function ` +
          `with \`observe(fn, '${node.name}')\`, or if already wrapped, ensure ` +
          `the name matches exactly.`
        );
      }
    }
  }

  return result;
}

/**
 * Load the most recent trace and check it against a DAG JSON file.
 *
 * Returns a TraceCheckResult with match details.
 */
export async function checkLastTrace(
  dagJsonPath: string
): Promise<TraceCheckResult> {
  const { getConfig } = await import("../config");
  const { ObservationStore } = await import("../storage/store");

  // Parse DAG
  const [dagNodes, parseErrors] = parseDag(dagJsonPath);
  if (parseErrors.length > 0) {
    return { valid: false, matched: [], unmatched: [], extraSpans: [], errors: parseErrors };
  }

  // Load latest trace
  const config = getConfig();
  const store = new ObservationStore(config.dbPath);
  try {
    store.createTables();
    const traces = store.listTraces(1);
    if (traces.length === 0) {
      return {
        valid: false,
        matched: [],
        unmatched: [],
        extraSpans: [],
        errors: ["No traces found. Run the app first to produce a trace."],
      };
    }

    const traceId = traces[0]["traceId"] as string;
    const tree = store.getTrace(traceId);
    if (tree.length === 0) {
      return {
        valid: false,
        matched: [],
        unmatched: [],
        extraSpans: [],
        errors: [`No spans found for trace '${traceId}'.`],
      };
    }

    return checkTraceAgainstDag(dagNodes, tree);
  } finally {
    store.close();
  }
}
